File: src/commands/audio-speech-azure.ts
import { Credentials } from '../credentials';
import { RefreshDeps } from '../refresh';
import { getAccessToken } from '../auth';
import {
  BinaryOutputDeps,
  ensureNotWritingBinaryToTTY,
  readStdinText,
  writeApiErrorAndExit,
  writeBinaryOutput,
} from '../audio-io';
import { ENDPOINTS } from '../endpoints';

export interface AudioSpeechAzureOptions {
  format?: string;
  output?: string;
}

export interface AudioSpeechAzureDeps extends BinaryOutputDeps {
  credentials?: Credentials;
  refreshDeps?: RefreshDeps;
  fetch?: typeof globalThis.fetch;
  now?: () => Date;
  stdin?: NodeJS.ReadableStream;
}

const DEFAULT_FORMAT = 'audio-24khz-48kbitrate-mono-mp3';

export async function audioSpeechAzureCommand(
  bodyArg: string | undefined,
  options: AudioSpeechAzureOptions,
  deps?: AudioSpeechAzureDeps,
): Promise<void> {
  const fetchFn = deps?.fetch ?? globalThis.fetch;
  const stderr = deps?.stderr ?? process.stderr;

  let body = bodyArg ?? '';
  if (!body || body === '-') {
    body = await readStdinText(deps?.stdin ?? process.stdin);
  }
  body = body.trim();
  if (!body) {
    throw new Error('No input. Pass an SSML document as an argument or pipe it via stdin.');
  }

  if (!body.startsWith('<speak')) {
    throw new Error(
      'Azure TTS requires SSML. Body must start with `<speak …>`. ' +
        'Tip: keep SSML in a file and pipe it in to avoid shell-escaping issues:\n' +
        '  monocle audio speech-azure -o out.mp3 < my.ssml',
    );
  }

  ensureNotWritingBinaryToTTY(options.output, deps ?? {});

  const { token, routerUrl } = await getAccessToken(deps);

  const response = await fetchFn(`${routerUrl}${ENDPOINTS.azureTextToSpeech}`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/ssml+xml',
      'X-Microsoft-OutputFormat': options.format ?? DEFAULT_FORMAT,
    },
    body,
  });

  if (!response.ok) await writeApiErrorAndExit(response as any, stderr);

  const buffer = Buffer.from(await response.arrayBuffer());
  writeBinaryOutput(buffer, options.output, deps ?? {});
}

File: src/commands/audio_speech_azure.rs
use crate::audio_io::{
    ensure_not_writing_binary_to_tty, read_stdin_text, write_api_error_and_exit,
    write_binary_output,
};
use crate::auth::get_access_token;
use crate::credentials::Credentials;
use crate::endpoints;
use crate::error::{AppError, Result};
use crate::net::Client;
use crate::origin::origin_header;

const DEFAULT_FORMAT: &str = "audio-24khz-48kbitrate-mono-mp3";

#[derive(Default)]
pub struct AudioSpeechAzureOptions {
    pub format: Option<String>,
    pub output: Option<String>,
}

pub fn audio_speech_azure_command(
    client: &Client,
    creds: &Credentials,
    body_arg: Option<&str>,
    options: AudioSpeechAzureOptions,
) -> Result<()> {
    let mut body = body_arg.unwrap_or("").to_string();
    if body.is_empty() || body == "-" {
        body = read_stdin_text()?;
    }
    let body = body.trim().to_string();
    if body.is_empty() {
        return Err(AppError::new(
            "No input. Pass an SSML document as an argument or pipe it via stdin.",
        ));
    }

    if !body.starts_with("<speak") {
        return Err(AppError::new(
            "Azure TTS requires SSML. Body must start with `<speak …>`. \
             Tip: keep SSML in a file and pipe it in to avoid shell-escaping issues:\n\
             \x20 monocle audio speech-azure -o out.mp3 < my.ssml",
        ));
    }

    ensure_not_writing_binary_to_tty(options.output.as_deref())?;

    let session = get_access_token(client, creds);

    let bearer = format!("Bearer {}", session.token);
    let format = options.format.as_deref().unwrap_or(DEFAULT_FORMAT);
    let resp = client.post_bytes(
        &format!("{}{}", session.router_url, endpoints::AZURE_TEXT_TO_SPEECH),
        &[
            ("Authorization", &bearer),
            ("X-Microsoft-OutputFormat", format),
            origin_header(),
        ],
        "application/ssml+xml",
        body.into_bytes(),
    )?;

    if !resp.ok() {
        write_api_error_and_exit(&resp);
    }

    write_binary_output(resp.bytes(), options.output.as_deref())
}
